ROS2: High CPU load of python scripts

Getting the following error:
`[ERROR] [ensenso_camera_node-10]: process has died [pid 17673, exit code -9`
when running the `ensenso_camera_node` with a virtual camera and virtual objects. It crashes after around ~300-350 seconds.

--- a/README.md
+++ b/README.md
@@ -5,7 +5,8 @@
 
 To get started with the package, take a look at the [ROS wiki](http://wiki.ros.org/ensenso_driver).
 
-If you want to use the package with ROS2, follow our [instructions](docs/ROS2.md) on how to build for ROS2.
+If you want to use the package with ROS2, see our [ROS2 documentation](docs/ROS2.md) on how to build for ROS2 and for
+further information.
 
 ## Remarks
  * For using this package, you need to have the [Ensenso SDK](https://www.ensenso.com/support/sdk-download) installed.

--- a/docs/ROS2.md
+++ b/docs/ROS2.md
@@ -64,6 +64,17 @@ ros2 launch ensenso_camera <tab><tab>
 
 ```
 
+## Documentation
+The usage is basically the same as for ROS1 except of course for the different ROS2 CLI as shown above.
+
+There is no documentation platform for third-party packages for ROS2 yet, but since we neither changed the API nor the
+message and action definitions for ROS2, you can simply refer to our [ROS1 wiki](http://wiki.ros.org/ensenso_driver).
+
+We still provide the same nodes as for ROS1, but one thing that has changed from ROS1 to ROS2 is that there are no
+nodelets in ROS2 (the concept has been replaced by that of a `Component`), which is why the launch file names have
+changed and some launch files have even been removed. Please keep this in mind when doing the
+[tutorials](http://wiki.ros.org/ensenso_driver/Tutorials) on nodelets.
+
 ## Limitations
 Since ROS2 does not support type masquerading (yet and probably never), point clouds in ROS2 are published as
 `sensor_msgs::msg::PointCloud2` and not directly as `pcl::PointCloud<T>` messages as in ROS1. This requires the user to

--- a/ensenso_camera/scripts/color_point_cloud
+++ b/ensenso_camera/scripts/color_point_cloud
@@ -64,6 +64,7 @@ def _main(node_name):
         texture_goal.far_plane = 4000.0
         texture_goal.near_plane = 100.0
 
+        node.get_logger().info("Texturing point cloud.")
         ros2py.send_action_goal(node, texture_point_cloud_client, texture_goal)
 
     ros2py.execute_at_rate(node, main_loop, loop_rate)

--- a/ensenso_camera/scripts/fit_primitive
+++ b/ensenso_camera/scripts/fit_primitive
@@ -139,6 +139,7 @@ def _main(node_name):
                 marker_array.markers.append(marker)
                 marker_id += 1
 
+        node.get_logger().info("Publishing marker array with {} elements.".format(len(marker_array.markers)))
         marker_publisher.publish(marker_array)
 
     ros2py.execute_at_rate(node, main_loop, loop_rate)

--- a/ensenso_camera/scripts/image_stream
+++ b/ensenso_camera/scripts/image_stream
@@ -61,6 +61,8 @@ def _main(node_name):
                 image = result.right_raw_images[0]
             else:
                 image = result.left_raw_images[0]
+
+        node.get_logger().info("Publishing image.")
         image_publisher.publish(image)
 
     ros2py.execute_at_rate(node, main_loop, loop_rate)

--- a/ensenso_camera/scripts/image_stream_mono
+++ b/ensenso_camera/scripts/image_stream_mono
@@ -45,6 +45,7 @@ def _main(node_name):
             image = result.rectified_images[0]
         else:
             image = result.raw_images[0]
+        node.get_logger().info("Publishing image.")
         image_publisher.publish(image)
 
     ros2py.execute_at_rate(node, main_loop, loop_rate)

--- a/ensenso_camera/scripts/request_data
+++ b/ensenso_camera/scripts/request_data
@@ -42,6 +42,8 @@ def _main(node_name):
         if result.error.code != 0:
             node.get_logger().error(ros2py.format_error(result.error))
 
+        node.get_logger().info("Data successfully requested.")
+
     ros2py.execute_at_rate(node, main_loop, loop_rate)
 
 

--- a/ensenso_camera/scripts/request_data_linked
+++ b/ensenso_camera/scripts/request_data_linked
@@ -51,7 +51,8 @@ def _main(node_name):
         if mono_result.error.code != 0:
             node.get_logger().error(ros2py.format_error(mono_result.error, "Error while acquiring mono data!"))
             return
-        pass
+
+        node.get_logger().info("Data successfully requested.")
 
     ros2py.execute_at_rate(node, main_loop, loop_rate)
 

--- a/ensenso_camera/scripts/request_data_mono
+++ b/ensenso_camera/scripts/request_data_mono
@@ -36,6 +36,8 @@ def _main(node_name):
         if result.error.code != 0:
             node.get_logger().error(ros2py.format_error(result.error))
 
+        node.get_logger().info("Data successfully requested.")
+
     ros2py.execute_at_rate(node, main_loop, loop_rate)
 
 

--- a/ensenso_camera/src/ensenso_camera/ros2.py
+++ b/ensenso_camera/src/ensenso_camera/ros2.py
@@ -40,6 +40,8 @@ if is_ros2():
 
     from action_msgs.msg import GoalStatus
     from rclpy.action import ActionClient
+    from rclpy.callback_groups import ReentrantCallbackGroup
+    from rclpy.executors import MultiThreadedExecutor
 
     def import_action(package_name, action_name):
         return import_from_module(package_name + ".action", action_name)
@@ -63,34 +65,36 @@ if is_ros2():
         def __init__(self, node, client, goal, feedback_callback=None):
             self._node = node
             self._client = client
+            self._event = threading.Event()
             self._goal = goal
             self._feedback_callback = feedback_callback
             self._response = None
+            self._result = None
 
         def send_goal(self):
-            self._goal_handle_future = self._client.send_goal_async(self._goal, self._feedback_callback)
+            self._event.clear()
+            send_goal_future = self._client.send_goal_async(self._goal, feedback_callback=self._feedback_callback)
+            send_goal_future.add_done_callback(self.response_callback)
 
         def wait_for_response(self, timeout_sec=None):
-            if self._node.spinning:
-                wait_until_future_complete(self._node, self._goal_handle_future)
-            else:
-                rclpy.spin_until_future_complete(self._node, self._goal_handle_future, timeout_sec=timeout_sec)
+            # Separate sending goal and waiting for response so that we can handle several clients at once by sending
+            # all the goals first and then waiting until all clients have received the response.
+            self._event.wait(timeout_sec)
+            self._response = ActionResponse(self._result.status, self._result.result)
 
-            goal_handle = self._goal_handle_future.result()
-
+        def response_callback(self, future):
+            goal_handle = future.result()
             # ROS2 does not provide means to check whether the retrieval of an action result timed out, but it provides
             # an `accepted()` method.
             if not goal_handle.accepted:
                 self._response = ActionResponse(GoalStatus.STATUS_ABORTED)
                 return
+            get_result_future = goal_handle.get_result_async()
+            get_result_future.add_done_callback(self.result_callback)
 
-            result_future = goal_handle.get_result_async()
-            if self._node.spinning:
-                wait_until_future_complete(self._node, result_future)
-            else:
-                rclpy.spin_until_future_complete(self._node, result_future, timeout_sec=timeout_sec)
-
-            self._response = ActionResponse(result_future.result().status, result_future.result().result)
+        def result_callback(self, future):
+            self._result = future.result()
+            self._event.set()
 
         def get_response(self):
             return self._response
@@ -103,10 +107,6 @@ if is_ros2():
         rate = node.create_rate(frequency)
         rate.sleep()
 
-    def wait_until_future_complete(node, future):
-        while not future.done():
-            sleep(node, 0.1)
-
     def wrap_main_function(main, node_name):
         try:
             main(node_name)
@@ -116,13 +116,12 @@ if is_ros2():
     def create_node(name, args=sys.argv):
         rclpy.init(args=args)
         node = rclpy.create_node(name)
-        node.thread = threading.Thread(target=rclpy.spin, args=(node,), daemon=True)
-        node.spinning = False
-        return node
-
-    def spin(node):
-        node.thread.start()
+        executor = MultiThreadedExecutor()
+        node.thread = threading.Thread(target=rclpy.spin, args=(node, executor), daemon=True)
+        node.callback_group = ReentrantCallbackGroup()
         node.spinning = True
+        node.thread.start()
+        return node
 
     def ok():
         return rclpy.ok()
@@ -180,7 +179,6 @@ if is_ros2():
         """
         action_handler = ActionHandler(node, client, goal, feedback_callback)
         action_handler.send_goal()
-
         action_handler.wait_for_response(timeout_sec)
         return action_handler.get_response()
 
@@ -202,14 +200,13 @@ if is_ros2():
         return responses
 
     def create_action_client(node, action_name, action):
-        return ActionClient(node, action, action_name)
+        return ActionClient(node, action, action_name, callback_group=node.callback_group)
 
     def execute_at_rate(node, func, rate_in_hz):
         # As explained in https://answers.ros.org/question/358343/, we have to spin because Rate in ROS2 registers a
         # callback with a ROS timer and without spinning the callback is never executed and thus the timer will never
         # wake up and trigger.
         try:
-            spin(node)
             rate = node.create_rate(rate_in_hz)
             while rclpy.ok():
                 func()

--- a/ensenso_camera/src/ensenso_camera/ros2.py
+++ b/ensenso_camera/src/ensenso_camera/ros2.py
@@ -145,11 +145,13 @@ if is_ros2():
         """
         Returns True if the client established a server connection in the given amount of time, False otherwise.
         """
+        node.get_logger().info("Connecting to action server {} ...".format(client._action_name))
         if not client.wait_for_server(timeout_sec):
             node.get_logger().error(SERVER_TIMEOUT_ERROR_MESSAGE)
             if exit:
                 sys.exit()
             return False
+        node.get_logger().info("Connected!")
         return True
 
     def wait_for_servers(node, clients, timeout_sec=None, exit=False):
@@ -335,11 +337,13 @@ else:
         Returns True if the client established a server connection in the given amount of time, False otherwise.
         """
         timeout = rospy.Duration() if timeout_sec is None else rospy.Duration(timeout_sec)
+        node.get_logger().info("Connecting to action server {} ...".format(client.action_client.ns))
         if not client.wait_for_server(timeout):
             node.get_logger().error(SERVER_TIMEOUT_ERROR_MESSAGE)
             if exit:
                 sys.exit()
             return False
+        node.get_logger().info("Connected!")
         return True
 
     def wait_for_servers(node, clients, timeout_sec=None, exit=True):

--- a/ensenso_camera2/include/ensenso_camera/queued_action_server.h
+++ b/ensenso_camera2/include/ensenso_camera/queued_action_server.h
@@ -42,11 +42,29 @@ public:
                      bool autoStart = false)
     : nh(nh), name(name), callback(callback)
   {
+    // -----------------------------------------------------------------------------------------------------------------
+    // Disable result caching by setting the timeout to 100ms. The default value of 15min caused the stereo_camera node
+    // to have a huge RAM usage when e.g. the request_data result included images.
+    // See: - https://design.ros2.org/articles/actions.html#result-caching
+    //      - https://answers.ros.org/question/366054
+    // -----------------------------------------------------------------------------------------------------------------
+    // BUG:
+    // Setting the timeout to value 0, which should discard a goal result immediately after responding to any pending
+    // result request as described in the linked article above, does not yet work with our code as expected. Some tests
+    // fail because the goal result status is UNKNOWN (0) and not SUCCEEDED (4). The current assumption is that the goal
+    // gets deleted before the result callback is executed.
+    // -----------------------------------------------------------------------------------------------------------------
+    // TODO:
+    // Find our whether this is a bug on our side or within the ROS2 action server implementation.
+    // -----------------------------------------------------------------------------------------------------------------
+    auto options = rcl_action_server_get_default_options();
+    options.result_timeout.nanoseconds = 1e6 * 100;  // 100ms
+
     this->actionServer = rclcpp_action::create_server<ActionT>(
         nh->get_base_interface(), nh->get_clock_interface(), nh->get_logging_interface(), nh->get_waitables_interface(),
         name, std::bind(&QueuedActionServer::onGoalReceived, this, std::placeholders::_1, std::placeholders::_2),
         std::bind(&QueuedActionServer::onCancelReceived, this, std::placeholders::_1),
-        std::bind(&QueuedActionServer::onGoalAccepted, this, std::placeholders::_1));
+        std::bind(&QueuedActionServer::onGoalAccepted, this, std::placeholders::_1), options);
 
     if (autoStart)
     {

--- a/ensenso_camera2/include/ensenso_camera/queued_action_server.h
+++ b/ensenso_camera2/include/ensenso_camera/queued_action_server.h
@@ -27,22 +27,15 @@ class QueuedActionServer
 {
 public:
   using ActionServerSharedPtr = typename rclcpp_action::Server<ActionT>::SharedPtr;
-
   using Goal = typename ActionT::Goal;
   using GoalConstSharedPtr = std::shared_ptr<const Goal>;
-
   using GoalHandle = typename rclcpp_action::ServerGoalHandle<ActionT>;
   using GoalHandleSharedPtr = std::shared_ptr<GoalHandle>;
-
+  using ExecuteCallback = std::function<void(GoalConstSharedPtr)>;
   using GoalUUID = rclcpp_action::GoalUUID;
-
-  using Result = typename ActionT::Result;
-
   using Feedback = typename ActionT::Feedback;
   using FeedbackConstSharedPtr = std::shared_ptr<const Feedback>;
-
-  // using ExecuteCallback = ensenso::std::function<void(GoalConstSharedPtr)>;
-  using ExecuteCallback = std::function<void(GoalConstSharedPtr)>;
+  using Result = typename ActionT::Result;
 
 public:
   QueuedActionServer(ensenso::ros::NodeHandle nh, std::string const& name, ExecuteCallback callback,

--- a/ensenso_camera_test/src/ensenso_camera_test/telecentric_projection.py
+++ b/ensenso_camera_test/src/ensenso_camera_test/telecentric_projection.py
@@ -62,10 +62,6 @@ class TestTelecentricProjection(unittest.TestCase):
         # Test the subscribed cloud
         self.send_goal_with_publishing_point_cloud()
 
-        # Explicitly spin the node so that the subscription callback will be executed, because the test cases do not
-        # send any further actions, which would implicitly spin the node
-        ros2py.spin(self.node)
-
     def tearDown(self):
         if os.path.isfile(IMAGE_PATH_MONO8):
             os.remove(IMAGE_PATH_MONO8)

--- a/ensenso_camera_test2/ensenso_camera_test/ros2_testing_hz.py
+++ b/ensenso_camera_test2/ensenso_camera_test/ros2_testing_hz.py
@@ -52,10 +52,8 @@ def test_func(self):
     measured_rates = HzTestSampler(self.test_name, self.msg_type, self.topic, self.params.duration_in_s).run()
     self.assertTrue(len(measured_rates) > 0, msg=f"No messages received on topic {self.topic}")
 
-    # TODO Find out why some of the GitHub action runs fail with very large rate deviations and reenable checking for
-    # the actual average rate.
-    # average_rate = sum(measured_rates) / len(measured_rates)
-    # self.assertAlmostEqual(average_rate, self.params.rate_in_hz, delta=self.params.tolerance_in_hz)
+    average_rate = sum(measured_rates) / len(measured_rates)
+    self.assertAlmostEqual(average_rate, self.params.rate_in_hz, delta=self.params.tolerance_in_hz)
 
 
 def HzTest(test_name, msg_type, topic, params):

--- a/ensenso_camera_test2/ensenso_camera_test/ros2_testing_hz.py
+++ b/ensenso_camera_test2/ensenso_camera_test/ros2_testing_hz.py
@@ -25,7 +25,6 @@ class HzTestSampler:
 
         self.node = ros2py.create_node(self.test_name)
         self.subscription = self.node.create_subscription(self.msg_type, self.topic, self.subscription_callback, 10)
-        ros2py.spin(self.node)
 
     def run(self):
         loop_rate = self.node.create_rate(0.1)

